pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -425,28 +443,33 @@ func (h *handler) EnqueueAfter(infra *infraObject, duration time.Duration) {
 	}
 }
 
+// OnChange is called whenever the infrastructure machine is updated, including when the object is being deleted.
 func (h *handler) OnChange(obj runtime.Object) (runtime.Object, error) {
 	infra, err := newInfraObject(obj)
 	if err != nil {
 		return obj, err
 	}
 
-	// don't process create if deleting
-	if !infra.meta.GetDeletionTimestamp().IsZero() {
-		return obj, nil
-	}
-
 	machine, err := rke2.GetOwnerCAPIMachine(obj, h.machineCache)
 	if apierrors.IsNotFound(err) {
 		logrus.Debugf("[machineprovision] %s/%s: waiting: machine to be set as owner reference", infra.meta.GetNamespace(), infra.meta.GetName())
 		h.EnqueueAfter(infra, 10*time.Second)
 		return obj, generic.ErrSkip
 	}
+
 	if err != nil {
-		logrus.Errorf("[machineprovision] %s/%s: error getting machine by owner reference %v", infra.meta.GetNamespace(), infra.meta.GetName(), err)
+		logrus.Errorf("[machineprovision] %s/%s: error getting machine by owner reference: %v", infra.meta.GetNamespace(), infra.meta.GetName(), err)
 		return obj, err
 	}
 
+	// the owner reference may be deleted if the machine is cleaned up forcefully
+	if infra.meta.GetLabels()[CapiMachineName] == "" {
+		infra.data.SetNested(machine.Name, "metadata", "labels", CapiMachineName)
+		return h.dynamic.Update(&unstructured.Unstructured{
```

Comment:
can we add a comment here stating that we don't want to "continue" logic if we run this Update because we could encounter an error further down and we're trying to keep the cache as up to date as possible